crates/typst/src/layout/grid/layout.rs

```diff
@@ -677,39 +954,155 @@ impl<'a> GridLayouter<'a> {
+                        // these index operations.
+                        y / 2
+                    })
+                    .map(|hlines| &**hlines)
+                    .unwrap_or(&[]);
+                let tracks = self.rcols.iter().copied().enumerate();
+
+                // Determine all different line segments we have to draw in
+                // this row, and convert them to points and shapes.
+                let segments = generate_line_segments(
+                    self.grid,
+                    tracks,
+                    y,
+                    hlines_at_row,
+                    is_bottom_border,
+                    hline_stroke_at_column,
+                )
+                .into_iter()
+                .map(|(stroke, dx, length)| {
+                    let stroke = (*stroke).clone().unwrap_or_default();
+                    let thickness = stroke.thickness;
+                    let half = thickness / 2.0;
+                    let dx = if self.is_rtl { self.width - dx - length } else { dx };
+                    let target = Point::with_x(length + thickness);
+                    let hline = Geometry::Line(target).stroked(stroke);
+                    (
+                        thickness,
+                        Point::new(dx - half, dy),
                         FrameItem::Shape(hline, self.span),
-                    );
-                }
+                    )
+                });
 
-                // Render vertical lines.
-                for (x, dx) in points(self.rcols.iter().copied()).enumerate() {
-                    let dx = if self.is_rtl { self.width - dx } else { dx };
-                    // We want each vline to span the entire table (start
-                    // at y = 0, end after all rows).
-                    // We use 'split_vline' to split the vline such that it
-                    // is not drawn above colspans.
-                    for (dy, length) in
-                        split_vline(self.grid, rows, x, 0, self.grid.rows.len())
-                    {
-                        let target = Point::with_y(length + thickness);
-                        let vline = Geometry::Line(target).stroked(stroke.clone());
-                        frame.prepend(
-                            Point::new(dx, dy - half),
-                            FrameItem::Shape(vline, self.span),
-                        );
-                    }
-                }
+                // Draw later (after we sort all lines below.)
+                lines.extend(segments);
+            }
+
+            // Sort by increasing thickness, so that we draw larger strokes
+            // on top.
+            // This avoids layering problems where a smaller hline appears
+            // "inside" a larger vline. When both have the same size, hlines
+            // are drawn on top (since the sort is stable, and they are
+            // pushed first).
+            lines.sort_by_key(|(thickness, ..)| *thickness);
```

Comment:
Worth mentioning that I considered using rayon's `par_sort_by` here. Should in theory be faster, but I wasn't sure. If that sounds interesting, I can run some unscientific benchmarks.

Reply:
The unscientific benchmarks have spoken: **using rayon's `par_sort_by_key` is slower.**

- 100 tables benchmark (modified to add different stroke sides for each cell): ~1.71s (no rayon) -> ~1.79s (with rayon)
- 10000 tables benchmark (same modification): ~61s (no rayon) -> ~63s (with rayon)